### DISCUSSION TITLE
RuntimeError: Requested backend tpu_driver, but it failed to initialize

### DISCUSSION THREAD

**Discussion by TMcMac** [jax-ml/jax#7787] [Q&A] [answered]
Hey Everyone,

So I am trying to work my way through a Jax tutorial and I am getting stuck when trying to replicate the very basics of jax.pmap(). The issue I seems to be hitting is that I am missing the driver for using TPUs. Here is the code block I am using:

import jax
import jax.numpy as jnp
impurt numpy as np
import requests
import os

if 'TPU_DRIVER_MODE' not in globals():
    url = 'http://' + os.environ['COLAB_TPU_ADDR'].split(':')[0] + ':8475/requestversion/tpu_driver_nightly'
    resp = requests.post(url)
    TPU_DRIVER_MODE = 1

from jax.config import config
config.FLAGS.jax_xla_backend = "tpu_driver"
config.FLAGS.jax_backend_target = "grpc://" + os.environ['COLAB_TPU_ADDR']
print(config.FLAGS.jax_backend_target)

Then I run jax.devices() and I get:

`RuntimeError: Requested backend tpu_driver, but it failed to initialize: Not found: Unable to find driver in registry given worker: ` ![Screen Shot 2021-09-02 at 9 54 49 AM](https://user-images.githubusercontent.com/15660681/131866651-b0879391-0c3d-4f8d-a98c-c292bf46589f.png)

I have the TPU selected from the hardware accelerators so it isn't that. Any ideas what I am missing?

![Screen Shot 2021-09-02 at 9 53 51 AM](https://user-images.githubusercontent.com/15660681/131866486-1de7bcdd-b27f-4e9d-8dd7-064a3dc1cf27.png)

**Comment by jakevdp** [comment, marked as answer]
Have you seen the `colab_tpu.setup_tpu()` utility? It takes care of most of this logic for you. Here is an example: https://jax.readthedocs.io/en/latest/jax-101/06-parallelism.html#colab-tpu-setup (or open directly [in Colab](https://colab.research.google.com/github/google/jax/blob/main/docs/jax-101/06-parallelism.ipynb#scrollTo=7mCgBzix2fd3))

I suspect the issue with your code has to do with using the `nightly` driver – sometimes that causes problems, which is why we generally pin the driver to a specific release that is known to work: https://github.com/google/jax/blob/f004bcb7b8763451b437202549faaea89f771f41/jax/tools/colab_tpu.py#L34-L37

**Reply by TMcMac** [reply]
Hey Jake, 

Thank you! The link you shared got me back on track with:
```
import jax

import jax.tools.colab_tpu

jax.tools.colab_tpu.setup_tpu()

jax.devices()
```
